hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/CaseExpression.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * Implements evaluation of SQL CASE operator.
+ */
+public final class CaseExpression<T> implements Expression<T>, IdentifiedDataSerializable {
+
+    private Expression<Boolean>[] conditions;
+    private Expression<?>[] results;
+    private QueryDataType resultType;
+
+    public CaseExpression() {
+        // No-op.
+    }
+
+    private CaseExpression(Expression<Boolean>[] conditions, Expression<?>[] results, QueryDataType resultType) {
+        this.conditions = conditions;
+        this.results = results;
+        this.resultType = resultType;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static CaseExpression<?> create(Expression<?>[] expressions, QueryDataType resultType) {
+        // Split conditions and expressions.
+        assert expressions != null;
+        assert expressions.length % 2 == 1;
```

Comment:
Why do we expect here uneven `expressions` count?

Reply:
clarified in 6201d4d8804a29f0b5cedbd5dad9d9a541788e2b